services/web/server/src/simcore_service_webserver/projects/_nodes_api.py

```diff
@@ -86,24 +108,113 @@ def guess_mimetype_if_undefined(cls, values):
+        if not f.file_id.endswith(".hidden_do_not_remove")
+        and Path(f.file_id).suffix.lower() in _SUPPORTED_PREVIEW_FILE_EXTENSIONS
+    }
+
+    with_thumbnail_image: list[_FileWithThumbnail] = []
+
+    for selected_file in set(selected_file_entries.keys()):
+        # search for thumbnail
+        thumbnail: FileMetaDataGet | None = None
+        for extension in _SUPPORTED_THUMBNAIL_EXTENSIONS:
+            thumbnail_search_file_name = f"{selected_file}{extension}"
+            if thumbnail_search_file_name in selected_file_entries:
+                thumbnail = selected_file_entries[thumbnail_search_file_name]
+                break
+        if not thumbnail:
+            continue
+
+        # since there is a thumbnail it can be associated to a file
+        with_thumbnail_image.append(
+            _FileWithThumbnail(
+                file=selected_file_entries[selected_file],
+                thumbnail=thumbnail,
+            )
+        )
+        # remove entries which have been used
+        selected_file_entries.pop(
+            __get_search_key(selected_file_entries[selected_file]), None
+        )
+        selected_file_entries.pop(__get_search_key(thumbnail), None)
+
+    no_thumbnail_image: list[_FileWithThumbnail] = [
+        _FileWithThumbnail(file=x, thumbnail=x) for x in selected_file_entries.values()
+    ]
+
+    return with_thumbnail_image + no_thumbnail_image
+
+
+async def __get_link(
+    app: web.Application, user_id: UserID, file_meta_data: FileMetaDataGet
+) -> tuple[str, HttpUrl]:
+    return __get_search_key(file_meta_data), await get_download_link(
+        app,
+        user_id,
+        parse_obj_as(SimCoreFileLink, {"store": "0", "path": file_meta_data.file_id}),
+    )
+
+
+async def _get_node_screenshots_parallel(
+    app: web.Application,
+    user_id: UserID,
+    files_with_thumbnails: list[_FileWithThumbnail],
 ) -> list[NodeScreenshot]:
-    """
-    ONLY for testing purposes
+    """resolves links in parallel before returning all NodeScreenshot"""
+
+    search_map: dict[str, FileMetaDataGet] = {}
+
+    for entry in files_with_thumbnails:
+        search_map[__get_search_key(entry.file)] = entry.file
+        search_map[__get_search_key(entry.thumbnail)] = entry.thumbnail
+
+    resolved_links: list[tuple[str, HttpUrl]] = await logged_gather(
+        *[__get_link(app, user_id, x) for x in search_map.values()],
+        max_concurrency=10,
+    )
+
+    mapped_http_url: dict[str, HttpUrl] = dict(resolved_links)
+
+    return [
+        NodeScreenshot(
+            mimetype=_guess_mimetype_from_name(e.file.file_id),
+            file_url=mapped_http_url[__get_search_key(e.file)],
+            thumbnail_url=mapped_http_url[__get_search_key(e.thumbnail)],
+        )
+        for e in files_with_thumbnails
+    ]
 
-    """
-    assert request.app[APP_SETTINGS_KEY].WEBSERVER_DEV_FEATURES_ENABLED  # nosec
-    screenshots = []
+
+async def get_node_screenshots(
+    request: web.Request,
```

Comment:
question: what is the point of passing the Request inside the api module?
I've seen this pattern already several times, I don't understand the use of separating handlers and apis modules if we pass the Request to the api. A request has to do with a handler, nothing with the api that only needs the Application.
So why not doing:

```suggestion
    app: web.application,
```